packages/react-storage/src/components/StorageBrowser/Views/Controls/Table.tsx

```diff
@@ -121,33 +81,129 @@ export interface TableControl<
 
 export type RenderRowItem<T> = (row: T, index: number) => JSX.Element;
 
+interface TableSortState<T> {
+  direction: SortDirection;
+  selection: keyof T;
+}
+
+type TableSortAction<T> = { selection: keyof T };
+
+export function tableSortReducer<T>(
+  state: TableSortState<T>,
+  action: TableSortAction<T>
+): TableSortState<T> {
+  const { direction, selection: prevSelection } = state;
+  const { selection } = action;
+
+  if (selection === prevSelection) {
+    const newDirection: SortDirection =
+      direction === 'ASCENDING' ? 'DESCENDING' : 'ASCENDING';
+
+    return { direction: newDirection, selection };
+  } else {
+    return { direction: 'ASCENDING', selection };
+  }
+}
+
+export function defaultTableSort<T>(
+  data: T[],
+  direction: SortDirection,
+  selection: keyof T
+): T[] {
+  return data.sort((a, b) => {
+    const isDateString = (s: T[keyof T]) =>
+      typeof s === 'string' && !isNaN(Date.parse(s));
+
+    const aValue = a[selection];
+    const bValue = b[selection];
+
+    // Handle missing values
+    if (aValue == null && bValue != null) {
+      return direction === 'ASCENDING' ? 1 : -1;
+    }
+    if (aValue != null && bValue == null) {
+      return direction === 'ASCENDING' ? -1 : 1;
+    }
+    if (aValue == null && bValue == null) {
+      return 0;
+    }
+
+    // If isDate is true, parse the values as Date objects
+    const aParsed =
+      typeof aValue === 'string' && isDateString(aValue)
+        ? new Date(aValue)
+        : aValue;
+    const bParsed =
+      typeof bValue === 'string' && isDateString(bValue)
+        ? new Date(bValue)
+        : bValue;
+
+    if (aParsed < bParsed) {
+      return direction === 'ASCENDING' ? -1 : 1;
+    } else if (aValue > bValue) {
+      return direction === 'ASCENDING' ? 1 : -1;
+    } else {
+      return 0;
+    }
+  });
+}
+
 interface TableControlProps<T> {
   data: T[];
   columns: Column<T>[];
   renderRowItem: RenderRowItem<T>;
+  sortState: TableSortState<T>;
+  updateTableSortState: (action: TableSortAction<T>) => void;
 }
 
 export const TableControl: TableControl = <U,>({
   data,
   columns,
   renderRowItem,
+  sortState,
+  updateTableSortState,
```

Comment:
At the moment the base `TableControl` is still handling the rendering for the table head and so I'm passing `sortState` to help display the sort icons. I also use `updateTableSortState` as the onClick handler for the table header button

Reply:
I can update to use a function similar to renderRowItem though

Reply:
hmm, I'm not 100% sure how I could create a `renderTableHeader` function similar to how `renderRowItem` works because the table headers need to know the sort direction and selection. If I'm creating a function that gets called inside the base table contro sot hat it renders the table hadersl, it would still need access to the sort state I think? 🤔